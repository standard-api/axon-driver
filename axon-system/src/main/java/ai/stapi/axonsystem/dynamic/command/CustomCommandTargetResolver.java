package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
