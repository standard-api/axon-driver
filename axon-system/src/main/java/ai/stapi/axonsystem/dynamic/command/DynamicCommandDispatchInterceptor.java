package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
