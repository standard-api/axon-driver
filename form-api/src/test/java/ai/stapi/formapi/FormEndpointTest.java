package ai.stapi.formapi;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.graphsystem.systemfixtures.model.SystemModelDefinitionsLoader;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.test.domain.DomainTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import java.util.List;
import java.util.Map;

import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@AutoConfigureMockMvc
@StructureDefinitionScope(SystemModelDefinitionsLoader.SCOPE)
class FormEndpointTest extends DomainTestCase {

    @Autowired
    private MockMvc mockMvc;

    @Test
    public void itShouldRespondWithJsonSchema() throws Exception {
        var result = this.mockMvc.perform(MockMvcRequestBuilders.post("/form/CreateStructureDefinition"))
            .andDo(print())
            .andExpect(status().isOk())
            .andReturn();

        var stringResponse = result.getResponse().getContentAsString();
        var mapResponse = new ObjectMapper().readValue(stringResponse, new TypeReference<Map<String, Object>>() {
        });
        this.thenObjectApproved(mapResponse);
    }


    @Test
    public void itShouldNotOmitExtensionsWhenOmitExtensionsIsExplicitlyFalse() throws Exception {
        var result = this.mockMvc.perform(
                MockMvcRequestBuilders
                    .post("/form/CreateStructureDefinition")
                    .content("{\"omitExtension\": \"false\"}")
                    .contentType("application/json")
            ).andDo(print())
            .andExpect(status().isOk())
            .andReturn();

        var stringResponse = result.getResponse().getContentAsString();
        var mapResponse = new ObjectMapper().readValue(
            stringResponse,
            new TypeReference<Map<String, Object>>() {
            }
        );
        this.thenObjectApproved(mapResponse);
    }

    @Test
    public void itShouldRespondWithJsonSchemaWithTargetIdentifier() throws Exception {
        var resourceId = "SomeId";
        var eventFactoryId = "EventFactoryId";
        this.givenCommandIsDispatched(
            new DynamicCommand(
                new UniqueIdentifier(resourceId),
                "CreateAggregateDefinition",
                Map.of(
                    "id", resourceId,
                    "name", "Example Aggregate",
                    "description", "Example Aggregate Description",
                    "structure", Map.of("id", "SomeResource"),
                    "command", List.of(
                        Map.of(
                            "creationalPolicy", "always",
                            "operation", Map.of("id", "SomeOperation"),
                            "eventFactory", List.of(
                                Map.of(
                                    "id", eventFactoryId,
                                    "event", Map.of("id", "SomeEventId")
                                )
                            )
                        )
                    )
                )
            )
        );
        var result = this.mockMvc.perform(
                MockMvcRequestBuilders
                    .post("/form/AddModificationOnAggregateDefinitionCommandEventFactory")
                    .content(
                        """
                            {
                                "resourceId": "%s",
                                "targets": {
                                    "eventFactoryId": "CommandHandlerDefinitionEventFactory/%s"
                                }
                            }
                        """.formatted(resourceId, eventFactoryId)
                    ).contentType("application/json")
            ).andDo(print())
            .andExpect(status().isOk())
            .andReturn();

        var stringResponse = result.getResponse().getContentAsString();
        var mapResponse = new ObjectMapper().readValue(
            stringResponse,
            new TypeReference<Map<String, Object>>() {
            }
        );
        this.thenObjectApproved(mapResponse);
    }

}