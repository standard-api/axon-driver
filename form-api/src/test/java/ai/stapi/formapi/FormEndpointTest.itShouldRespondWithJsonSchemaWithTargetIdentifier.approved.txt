{
  "formData" : { },
  "formSchema" : {
    "description" : "Generated command for adding modification(CommandHandlerDefinitionEventFactoryModification) on AggregateDefinition Aggregate",
    "title" : "AddModificationOnAggregateDefinitionCommandEventFactory",
    "type" : "object",
    "definitions" : {
      "CommandHandlerDefinitionEventFactoryModification" : {
        "description" : "Type for anonymous field contained in CommandHandlerDefinition",
        "title" : "Modification",
        "type" : "object",
        "properties" : {
          "id" : {
            "description" : "Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.",
            "title" : "Id",
            "type" : "string"
          },
          "startIdParameterName" : {
            "description" : "The parameter name on operation which will be used as starting node for modification.",
            "title" : "Start Id Parameter Name",
            "type" : "string"
          },
          "destinationIndexParameterName" : {
            "description" : "The parameter name on operation which will be used as destination index for insert and move",
            "title" : "Destination Index Parameter Name",
            "type" : "string"
          },
          "inputValueParameterName" : {
            "description" : "The parameter name on operation which will be used as input value for modification.",
            "title" : "Input Value Parameter Name",
            "type" : "string"
          },
          "kind" : {
            "description" : "The kind of modification to be made. add | replace | move | remove | insert",
            "title" : "Kind",
            "type" : "string"
          },
          "modificationPath" : {
            "description" : "The path on aggregate structure where the modification will be applied.",
            "title" : "Modification Path",
            "type" : "string"
          }
        },
        "required" : [ "kind", "modificationPath" ]
      }
    },
    "properties" : {
      "eventFactoryId" : {
        "description" : "",
        "title" : "Event Factory Id",
        "type" : "string"
      },
      "modification" : {
        "type" : "array",
        "items" : {
          "$ref" : "#/definitions/CommandHandlerDefinitionEventFactoryModification",
          "description" : "",
          "title" : "Modification"
        }
      }
    },
    "required" : [ "eventFactoryId", "modification" ]
  },
  "uiSchema" : { }
}